client/state/utils.js

```diff
@@ -23,6 +23,84 @@ export function isValidStateWithSchema( state, schema ) {
 }
 
 /**
+ * Creates a super-reducer as a map of reducers over keyed objects
+ *
+ * Use this when wanting to write reducers that operate
+ * on a single object as if it lived in isolation when
+ * really it lives in a map of similar objects referenced
+ * by some predesignated key or id. This could be used
+ * for example when reducing properties on a site object
+ * wherein we have many sites keyed by site id.
+ *
+ * Note! This will only apply the supplied reducer to
+ * the item referenced by the supplied key in the action.
+ *
+ * If no key exists whose name matches the given keyName
+ * then this super-reducer will abort and return the
+ * previous state.
+ *
+ * If some action should apply to every single item
+ * in the map of keyed objects, this utility cannot be
+ * used as it will only reduce the referenced item.
+ *
+ * @example
+ * const age = ( state = 0, action ) =>
+ *     GROW === action.type
+ *         ? state + 1
+ *         : state
+ *
+ * const title = ( state = 'grunt', action ) =>
+ *     PROMOTION === action.type
+ *         ? action.title
+ *         : state
+ *
+ * const userReducer = combineReducers( {
+ *     age,
+ *     title,
+ * } )
+ *
+ * export default keyedReducer( 'username', userReducer )
+ *
+ * dispatch( { type: GROW, username: 'hunter02' } )
+ *
+ * state.users === {
+ *     hunter02: {
+ *         age: 1,
+ *         title: 'grunt',
+ *     }
+ * }
+ *
+ * @param {string} keyName name of key in action referencing item in state map
+ * @param {function} reducer applied to referenced item in state map
+ * @return {function} super-reducer applying reducer over map of keyed items
+ */
+export const keyedReducer = ( keyName, reducer ) => ( state = {}, action ) => {
+	// the action must refer to some item in the map
+	const itemKey = action[ keyName ];
+
+	// if no reference exists abort
+	// and return unchanged state
+	if ( ! itemKey ) {
+		return state;
+	}
+
+	// pass the old sub-state from that item into the reducer
+	const oldItemState = state[ itemKey ];
+	const newItemState = reducer( oldItemState, action );
+
+	// and do nothing if the new sub-state matches the old sub-state
+	if ( newItemState === oldItemState ) {
```

Comment:
>For example isRequesting state

Taking that example, we'd assume that in calling a selector `isRequesting( state, 2916284 );`, it would return `false` if no requests had ever been made (i.e. initial state).